File: src/lib/classify.ts
// ─────────────────────────────────────────────────────────────────────────────
// classify.ts  —  Port of CELL 4 classification block (M3.4)
//
// Three-layer page classifier + corpus-aware Layer 3 inheritance + dual-mechanism
// hub detection + programmatic-series flagging. Hard rules preserved verbatim:
//   • Layer 3 must NOT inherit from the domain root (len(segs) < 2 → no parent).
//   • Hub detection is dual: data-driven (detect_hub_pages) AND vocabulary-driven
//     (_is_section_hub_url, applied in the Level 0 crawl filter).
//   • Programmatic series children are KEPT (re-tagged 'Other'), never dropped.
//
// FIX (M3.4.1): Added shouldExcludePage() — a single unified exclusion gate that
// is now called in BOTH crawl.ts (Level 0) and quickmatch.ts (working filter).
// Previously the Level 0 crawl filter and the quickmatch Hub-only filter were
// inconsistent: pages classified as Homepage/Contact, or matching SKIP_URL_PATTERNS
// (e.g. /category/*, /tag/*), were excluded from the metadata crawl but still
// reached quickmatch via their query_all surface. shouldExcludePage() closes that
// gap by centralising all exclusion logic in one place.
// ─────────────────────────────────────────────────────────────────────────────
import {
  CATEGORY_PATTERNS, CATEGORY_VOCAB, LAYER2_MIN_VOCAB_HITS, PROGRAMMATIC_SERIES_MIN,
} from "./config";
import { PageRow, LogFn, urlPath } from "./util";

const REAL_CATEGORIES = new Set(["Blog", "Landing Page", "Product / Service", "Contact"]);

// ── Layer 1 + Layer 2 ───────────────────────────────────────────────────────
export function classifyPage(url: string): string {
  let path: string;
  try {
    // Strip hash fragment — never part of server-side page identity (M3.4d).
    path = urlPath(url);
  } catch {
    return "Other";
  }
  if (path === "" || path === "/") return "Homepage";

  // Layer 1 — structural pattern match (exact, fast).
  for (const category of Object.keys(CATEGORY_PATTERNS)) {
    if (CATEGORY_PATTERNS[category].some((p) => path.includes(p))) return category;
  }

  // Layer 2 — slug vocabulary scoring (requires confident multi-token match).
  const tokens = new Set(path.split(/[-_/]/).filter(Boolean));
  let bestCat = "Other";
  let bestHits = 0;
  for (const category of Object.keys(CATEGORY_VOCAB)) {
    let hits = 0;
    for (const t of tokens) if (CATEGORY_VOCAB[category].has(t)) hits++;
    if (hits > bestHits) {
      bestCat = category;
      bestHits = hits;
    }
  }
  return bestHits >= LAYER2_MIN_VOCAB_HITS ? bestCat : "Other";
}

// ── Unified exclusion gate (M3.4.1) ─────────────────────────────────────────
// Returns true when a page must be dropped before scoring or crawling.
// Applied in BOTH crawl.ts (Level 0 filter) AND quickmatch.ts (working filter)
// so the two stages are always consistent.
//
// Exclusion criteria (any one is sufficient):
//   1. Category is Homepage, Contact, or Hub — not placeable content.
//   2. Path matches SKIP_URL_PATTERNS — structural/utility URLs that are never
//      placement candidates (taxonomy archives, auth pages, etc.).
//   3. Path is a bare section root — a URL whose path is a single well-known
//      section slug with no child segment (e.g. /blog/, /resources/). These are
//      listing/hub pages even when classifyPage() assigns them a content category
//      (e.g. /blog/ → "Blog") because detectHubPages() only fires when the domain
//      has enough child posts in the GSC date window.

// Taxonomy / utility URL patterns — extended to cover WordPress-style archives.
// Matches anywhere in the full path so /category/anything/ and /tag/foo/bar/ are
// both caught.  The blog-root pattern catches /blog and /blog/ as a bare root
// (terminal segment equals "blog") via BARE_SECTION_ROOTS below, so we no longer
// need the `\/blog\/?$` anchor here.
const SKIP_URL_PATTERNS =
  /\/(contact|privacy|terms|login|signin|cart|checkout|category|tag|author|page|feed|wp-json|wp-admin|wp-content|amp|search)\/?/i;

// Well-known section-root slugs. A path whose ONLY segment is one of these is a
// listing page (e.g. /blog/, /resources/, /guides/) and must be excluded even if
// classifyPage() returns a content category for it.
const BARE_SECTION_ROOTS = new Set([
  "blog", "blogs", "news", "insights", "resources", "resource",
  "articles", "article", "posts", "post", "updates", "update",
  "guides", "guide", "tips", "tip", "learn", "library",
  "press", "events", "event", "podcast", "podcasts", "webinars", "webinar",
  "case-studies", "case-study", "testimonials", "stories", "story",
]);

export function shouldExcludePage(url: string, page_category: string): boolean {
  // 1. Category-based exclusion.
  if (["Homepage", "Contact", "Hub"].includes(page_category)) return true;

  let path: string;
  try {
    path = new URL(url.split("#")[0]).pathname.replace(/\/+$/, "").toLowerCase();
  } catch {
    return true; // unparseable URL — exclude
  }

  // 2. Structural / utility URL pattern match.
  if (SKIP_URL_PATTERNS.test(path)) return true;

  // 3. Bare section root — path has exactly one non-empty segment that is a
  //    known listing slug. Catches /blog, /blog/, /resources/, /guides/ etc.
  const segs = path.split("/").filter(Boolean);
  if (segs.length === 1 && BARE_SECTION_ROOTS.has(segs[0])) return true;

  return false;
}

// (domain, parentPath) tuple key, or null for top-level pages (len(segs) < 2).
function parentKey(url: string): string | null {
  try {
    const parsed = new URL(url.split("#")[0]);
    const dom = parsed.host.replace(/^www\./, "");
    const segs = parsed.pathname.split("/").filter(Boolean);
    if (segs.length < 2) return null; // M3.4d: top-level pages never inherit.
    return `${dom}|${segs.slice(0, -1).join("/")}`;
  } catch {
    return null;
  }
}

function selfKey(url: string): string | null {
  try {
    const parsed = new URL(url.split("#")[0]);
    const dom = parsed.host.replace(/^www\./, "");
    const segs = parsed.pathname.split("/").filter(Boolean);
    return segs.length ? `${dom}|${segs.join("/")}` : null;
  } catch {
    return null;
  }
}

// ── Layer 3 — corpus-aware parent-path inheritance ──────────────────────────
export function enrichPageCategories(pages: PageRow[], log: LogFn = console.log): PageRow[] {
  if (!pages.length) return pages;

  // Dominant real category per named parent path.
  const byParent = new Map<string, Map<string, number>>();
  for (const p of pages) {
    const k = parentKey(p.page);
    if (k === null || !REAL_CATEGORIES.has(p.page_category)) continue;
    if (!byParent.has(k)) byParent.set(k, new Map());
    const m = byParent.get(k)!;
    m.set(p.page_category, (m.get(p.page_category) ?? 0) + 1);
  }
  const parentCat = new Map<string, string>();
  for (const [k, m] of byParent) {
    let best = "";
    let bestN = -1;
    for (const [cat, n] of m) if (n > bestN) { best = cat; bestN = n; }
    parentCat.set(k, best);
  }

  const before = pages.filter((p) => p.page_category === "Other").length;
  for (const p of pages) {
    if (p.page_category !== "Other") continue;
    const k = parentKey(p.page);
    if (k !== null && parentCat.has(k)) p.page_category = parentCat.get(k)!;
  }
  const after = pages.filter((p) => p.page_category === "Other").length;
  if (before > after) log(`▸ Category enrichment: ${before - after} 'Other' pages inherited a parent category.`);
  return pages;
}

// ── Hub detection (data-driven mechanism #1) ────────────────────────────────
export function detectHubPages(pages: PageRow[], log: LogFn = console.log): PageRow[] {
  if (!pages.length) return pages;
  const parentSet = new Set<string>();
  for (const p of pages) {
    const k = parentKey(p.page);
    if (k !== null) parentSet.add(k);
  }
  let n = 0;
  for (const p of pages) {
    const sk = selfKey(p.page);
    const isHub = sk !== null && parentSet.has(sk);
    // Do not override Contact / Other (Programmatic) flags already set.
    if (isHub && p.page_category !== "Contact" && p.page_category !== "Other") {
      p.page_category = "Hub";
      n++;
    }
  }
  if (n > 0) log(`▸ Hub pages: ${n} section index pages flagged (links to other pages, not standalone content).`);
  return pages;
}

// ── Programmatic series exclusion (KEPT, not dropped) ───────────────────────
export function detectProgrammaticSeries(pages: PageRow[], log: LogFn = console.log): PageRow[] {
  if (!pages.length) return pages;
  const counts = new Map<string, number>();
  for (const p of pages) {
    if (p.page_category !== "Other") continue;
    const k = parentKey(p.page);
    if (k !== null) counts.set(k, (counts.get(k) ?? 0) + 1);
  }
  const flagged = new Map<string, number>();
  for (const [k, c] of counts) if (c >= PROGRAMMATIC_SERIES_MIN) flagged.set(k, c);
  if (!flagged.size) return pages;

  let n = 0;
  for (const p of pages) {
    if (p.page_category !== "Other") continue;
    const k = parentKey(p.page);
    if (k !== null && flagged.has(k)) {
      p.page_category = "Other"; // surfaced for review with no special label
      n++;
    }
  }
  log(`▸ Programmatic series: ${flagged.size} pattern(s), ${n} subfolder children re-tagged as 'Other'.`);
  const top = [...flagged.entries()].sort((a, b) => b[1] - a[1]).slice(0, 8);
  for (const [k, c] of top) {
    const [dom, parent] = k.split("|");
    log(`   • /${parent}/ on ${dom} (${c} pages)`);
  }
  return pages;
}

File: backend/lib/classify.py
"""
classify.py — Page classifier + M3.5 content-type classifier.
Port of classify.ts. Hard rules preserved:
  D3: Layer 3 must NOT inherit from domain root (segs < 2 → no parent).
  D8: Programmatic series children are KEPT (re-tagged 'Other'), never dropped.
"""
from __future__ import annotations
import re
from typing import Optional

from .config import (
    CATEGORY_PATTERNS, CATEGORY_VOCAB, LAYER2_MIN_VOCAB_HITS, PROGRAMMATIC_SERIES_MIN,
)
from .util import PageRow, LogFn, url_path

REAL_CATEGORIES = {"Blog", "Landing Page", "Product / Service", "Contact"}


# ── Layer 1 + Layer 2 ─────────────────────────────────────────────────────────

def classify_page(url: str) -> str:
    try:
        path = url_path(url)
    except Exception:
        return "Other"
    if path == "" or path == "/":
        return "Homepage"

    # Layer 1 — structural pattern match
    for category, patterns in CATEGORY_PATTERNS.items():
        if any(p in path for p in patterns):
            return category

    # Layer 2 — slug vocabulary scoring
    tokens = set(re.split(r"[-_/]", path))
    best_cat = "Other"
    best_hits = 0
    for category, vocab in CATEGORY_VOCAB.items():
        hits = sum(1 for t in tokens if t in vocab)
        if hits > best_hits:
            best_cat = category
            best_hits = hits
    return best_cat if best_hits >= LAYER2_MIN_VOCAB_HITS else "Other"


def _parent_key(url: str) -> Optional[str]:
    """(domain, parentPath) key, or None for top-level pages (D3)."""
    try:
        from urllib.parse import urlparse
        parsed = urlparse(url.split("#")[0])
        dom = parsed.netloc.removeprefix("www.")
        segs = [s for s in parsed.path.split("/") if s]
        if len(segs) < 2:
            return None  # D3: top-level pages never inherit
        return f"{dom}|{'/'.join(segs[:-1])}"
    except Exception:
        return None


def _self_key(url: str) -> Optional[str]:
    try:
        from urllib.parse import urlparse
        parsed = urlparse(url.split("#")[0])
        dom = parsed.netloc.removeprefix("www.")
        segs = [s for s in parsed.path.split("/") if s]
        return f"{dom}|{'/'.join(segs)}" if segs else None
    except Exception:
        return None


# ── Layer 3 — corpus-aware parent-path inheritance ────────────────────────────

def enrich_page_categories(pages: list[PageRow], log: LogFn = print) -> list[PageRow]:
    if not pages:
        return pages

    by_parent: dict[str, dict[str, int]] = {}
    for p in pages:
        k = _parent_key(p.page)
        if k is None or p.page_category not in REAL_CATEGORIES:
            continue
        if k not in by_parent:
            by_parent[k] = {}
        by_parent[k][p.page_category] = by_parent[k].get(p.page_category, 0) + 1

    parent_cat: dict[str, str] = {}
    for k, m in by_parent.items():
        best = max(m, key=lambda c: m[c])
        parent_cat[k] = best

    before = sum(1 for p in pages if p.page_category == "Other")
    for p in pages:
        if p.page_category != "Other":
            continue
        k = _parent_key(p.page)
        if k is not None and k in parent_cat:
            p.page_category = parent_cat[k]
    after = sum(1 for p in pages if p.page_category == "Other")
    if before > after:
        log(f"▸ Category enrichment: {before - after} 'Other' pages inherited a parent category.")
    return pages


# ── Hub detection ─────────────────────────────────────────────────────────────

def detect_hub_pages(pages: list[PageRow], log: LogFn = print) -> list[PageRow]:
    if not pages:
        return pages
    parent_set: set[str] = set()
    for p in pages:
        k = _parent_key(p.page)
        if k is not None:
            parent_set.add(k)
    n = 0
    for p in pages:
        sk = _self_key(p.page)
        is_hub = sk is not None and sk in parent_set
        if is_hub and p.page_category not in ("Contact", "Other"):
            p.page_category = "Hub"
            n += 1
    if n > 0:
        log(f"▸ Hub pages: {n} section index pages flagged.")
    return pages


# ── Programmatic series (D8 — KEPT not dropped) ───────────────────────────────

def detect_programmatic_series(pages: list[PageRow], log: LogFn = print) -> list[PageRow]:
    if not pages:
        return pages
    counts: dict[str, int] = {}
    for p in pages:
        if p.page_category != "Other":
            continue
        k = _parent_key(p.page)
        if k is not None:
            counts[k] = counts.get(k, 0) + 1

    flagged = {k: c for k, c in counts.items() if c >= PROGRAMMATIC_SERIES_MIN}
    if not flagged:
        return pages

    n = 0
    for p in pages:
        if p.page_category != "Other":
            continue
        k = _parent_key(p.page)
        if k is not None and k in flagged:
            p.page_category = "Other"  # surfaced for review
            n += 1
    log(f"▸ Programmatic series: {len(flagged)} pattern(s), {n} subfolder children re-tagged as 'Other'.")
    top = sorted(flagged.items(), key=lambda x: -x[1])[:8]
    for k, c in top:
        dom, parent = k.split("|", 1)
        log(f"   • /{parent}/ on {dom} ({c} pages)")
    return pages


# ── M3.5: Content-type classifier ────────────────────────────────────────────
# Runs AFTER metadata crawl (needs H1/H2). Detects Listicle, How-to, Comparison.
# Priority order: Comparison > Listicle > How-to (Comparison is most specific).

_LISTICLE_TITLE_WORDS = {"best", "top", "ultimate", "list", "picks"}
_LISTICLE_SLUG_PATTERNS = re.compile(r"(?:^|[-/])(?:best|top|ultimate)[-/]|[-/]list(?:[-/]|$)")
_HOWTO_SLUG_PATTERNS = re.compile(r"(?:^|[-/])how[-_]to[-/]|[-/](?:guide|tutorial|steps?)(?:[-/]|$)")
_HOWTO_TITLE_WORDS = {"how", "guide", "tutorial", "step", "steps", "learn"}
_COMPARISON_SLUG_PATTERNS = re.compile(r"[-/]vs[-/]|versus|[-/]alternatives?(?:[-/]|$)|[-/]review(?:s?)(?:[-/]|$)")
_COMPARISON_TITLE_WORDS = {"vs", "versus", "alternatives", "alternative", "compared", "comparison", "review", "reviews"}
_NUMBERED_H2 = re.compile(r"(?:^|\| )\d+[\.\)]\s")  # "1. Thing | 2. Other"


def classify_content_type(page: PageRow) -> Optional[str]:
    """
    Classify a page as Listicle, How-to, or Comparison.
    Returns None if no type matches.
    Priority: Comparison > Listicle > How-to.
    """
    try:
        from urllib.parse import urlparse
        path = urlparse(page.page.split("#")[0]).path.lower()
    except Exception:
        path = ""

    title = (page.meta_title or "").lower()
    h1 = (page.h1 or "").lower()
    h2 = (page.h2 or "").lower()
    query = (page.query or "").lower()

    # ── Comparison (most specific — check first) ──────────────────────────────
    if (
        _COMPARISON_SLUG_PATTERNS.search(path)
        or any(w in title.split() for w in _COMPARISON_TITLE_WORDS)
        or any(w in h1.split() for w in _COMPARISON_TITLE_WORDS)
        or "vs" in query or "versus" in query or "alternative" in query
    ):
        return "Comparison"

    # ── Listicle ──────────────────────────────────────────────────────────────
    title_words = set(title.split())
    h1_words = set(h1.split())
    has_listicle_title = bool(title_words & _LISTICLE_TITLE_WORDS or h1_words & _LISTICLE_TITLE_WORDS)
    has_numbered_h2s = bool(_NUMBERED_H2.search(h2))
    has_listicle_slug = bool(_LISTICLE_SLUG_PATTERNS.search(path))
    if has_listicle_slug or (has_listicle_title and has_numbered_h2s):
        return "Listicle"

    # ── How-to ────────────────────────────────────────────────────────────────
    title_words_set = set(title.split())
    h1_words_set = set(h1.split())
    has_howto_title = bool(
        title_words_set & _HOWTO_TITLE_WORDS or h1_words_set & _HOWTO_TITLE_WORDS
    )
    has_howto_slug = bool(_HOWTO_SLUG_PATTERNS.search(path))
    has_step_h2s = bool(re.search(r"(?:^|\| )(?:step|stage)\s+\d+", h2))
    if has_howto_slug or (has_howto_title and (has_step_h2s or has_numbered_h2s)):
        return "How-to"

    return None


def enrich_content_types(pages: list[PageRow], log: LogFn = print) -> list[PageRow]:
    """Assign content_type to all pages. Called after metadata crawl."""
    n = 0
    for p in pages:
        ct = classify_content_type(p)
        p.content_type = ct
        if ct:
            n += 1
    if n:
        log(f"▸ Content types: {n} pages classified (Listicle / How-to / Comparison).")
    return pages
